File: src/main/java/me/miki/shindo/management/settings/impl/ComboSetting.java
package me.miki.shindo.management.settings.impl;

import me.miki.shindo.management.settings.config.ConfigOwner;
import me.miki.shindo.management.language.TranslateText;
import me.miki.shindo.management.settings.Setting;
import me.miki.shindo.management.settings.impl.combo.Option;

import java.util.ArrayList;
import java.util.List;

public class ComboSetting extends Setting {

    private final List<Option> options;
    private final Option defaultOption;
    private Option option;

    public ComboSetting(TranslateText text, ConfigOwner parent, TranslateText defaultOption, List<Option> options) {
        super(text, parent);
        this.options = new ArrayList<>(options);
        this.option = getOptionByKey(defaultOption.getKey());
        this.defaultOption = this.option;
    }

    public ComboSetting(TranslateText text, ConfigOwner parent, String defaultOptionKey, List<Option> options) {
        super(text, parent);
        this.options = new ArrayList<>(options);
        this.option = getOptionByKey(defaultOptionKey);
        if (this.option == null && !this.options.isEmpty()) {
            this.option = this.options.get(0);
        }
        this.defaultOption = this.option;
    }

    public ComboSetting(String name, ConfigOwner parent, String defaultOptionKey, List<Option> options) {
        super(name, parent);
        this.options = new ArrayList<>(options);
        this.option = getOptionByKey(defaultOptionKey);
        if (this.option == null && !this.options.isEmpty()) {
            this.option = this.options.get(0);
        }
        this.defaultOption = this.option;
    }

    @Override
    public void reset() {
        this.option = defaultOption;
    }

    public Option getOption() {
        return option;
    }

    public void setOption(Option option) {
        this.option = option;
    }

    public List<Option> getOptions() {
        return options;
    }

    public Option getDefaultOption() {
        return defaultOption;
    }

    @Deprecated
    public Option getOptionByNameKey(String nameKey) {
        return getOptionByKey(nameKey);
    }

    public Option getOptionByKey(String key) {
        for (Option op : options) {
            if (op.getNameKey().equalsIgnoreCase(key)) {
                return op;
            }
        }
        return options.isEmpty() ? null : options.get(0);
    }
}

File: src/main/kotlin/me/miki/shindo/management/settings/impl/ComboSetting.kt
package me.miki.shindo.management.settings.impl

import me.miki.shindo.management.language.TranslateText
import me.miki.shindo.management.settings.Setting
import me.miki.shindo.management.settings.config.ConfigOwner
import me.miki.shindo.management.settings.impl.combo.Option

open class ComboSetting : Setting {

    private val options: MutableList<Option>
    private val defaultOption: Option?
    private var option: Option?

    constructor(text: TranslateText, parent: ConfigOwner, defaultOption: TranslateText, options: List<Option>) : super(
        text,
        parent
    ) {
        this.options = ArrayList(options)
        this.option = getOptionByKey(defaultOption.getKey())
        this.defaultOption = this.option
    }

    constructor(text: TranslateText, parent: ConfigOwner, defaultOptionKey: String, options: List<Option>) : super(
        text,
        parent
    ) {
        this.options = ArrayList(options)
        this.option = getOptionByKey(defaultOptionKey)
        if (this.option == null && this.options.isNotEmpty()) {
            this.option = this.options[0]
        }
        this.defaultOption = this.option
    }

    constructor(name: String, parent: ConfigOwner, defaultOptionKey: String, options: List<Option>) : super(
        name,
        parent
    ) {
        this.options = ArrayList(options)
        this.option = getOptionByKey(defaultOptionKey)
        if (this.option == null && this.options.isNotEmpty()) {
            this.option = this.options[0]
        }
        this.defaultOption = this.option
    }

    override fun reset() {
        option = defaultOption
    }

    fun getOption(): Option? {
        return option
    }

    open fun setOption(option: Option?) {
        this.option = option
    }

    fun getOptions(): List<Option> {
        return options
    }

    fun getDefaultOption(): Option? {
        return defaultOption
    }

    fun getOptionByNameKey(nameKey: String): Option? {
        return getOptionByKey(nameKey)
    }

    private fun getOptionByKey(key: String): Option? {
        for (op in options) {
            if (op.nameKey.equals(key, ignoreCase = true)) {
                return op
            }
        }
        return if (options.isEmpty()) null else options[0]
    }
}
